go/ops_test.go

```diff
@@ -61,3 +64,160 @@ func TestDoHash(t *testing.T) {
 		})
 	}
 }
+
+func TestInnerOpCheckAgainstSpec(t *testing.T) {
+	var (
+		spec    *ProofSpec
+		innerOp *InnerOp
+	)
+
+	cases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			"success",
+			func() {},
+			nil,
+		},
+		{
+			"failure: empty spec",
+			func() {
+				spec = nil
+			},
+			errors.New("op and spec must be both non-nil"),
+		},
+		{
+			"failure: empty inner spec",
+			func() {
+				spec.InnerSpec = nil
+			},
+			errors.New("spec.InnerSpec must be non-nil"),
+		},
+		{
+			"failure: empty leaf spec",
+			func() {
+				spec.LeafSpec = nil
+			},
+			errors.New("spec.LeafSpec must be non-nil"),
+		},
+		{
+			"failure: incorrect hash function inner op",
+			func() {
+				innerOp.Hash = HashOp_BITCOIN
+			},
+			fmt.Errorf("unexpected HashOp: %d", HashOp_BITCOIN),
+		},
+		{
+			"failure: spec fails validation",
+			func() {
+				innerOp.Prefix = []byte{0x01}
+			},
+			fmt.Errorf("wrong value in IAVL leaf op"),
+		},
+		{
+			"failure: inner prefix starts with leaf prefix",
+			func() {
+				// change spec to be non-iavl spec to skip strict iavl validation
+				spec.LeafSpec.PrehashKey = HashOp_BITCOIN
+				innerOp.Prefix = append(spec.LeafSpec.Prefix, innerOp.Prefix...)
+			},
+			fmt.Errorf("inner Prefix starts with %X", []byte{0}),
+		},
+		{
+			"failure: inner prefix too short",
+			func() {
+				// change spec to be non-iavl spec to skip strict iavl validation
+				spec.LeafSpec.PrehashKey = HashOp_BITCOIN
+				innerOp.Prefix = []byte{0x01}
+			},
+			errors.New("innerOp prefix too short (1)"),
+		},
+		{
+			"failure: inner prefix too long",
+			func() {
+				// change spec to be non-iavl spec to skip strict iavl validation
+				spec.LeafSpec.PrehashKey = HashOp_BITCOIN
+				innerOp.Prefix = []byte("AgQIIGe3bHuC1g6+5/Qd0RoCU0waFu+nDCFzEDViMN/VrQwgIA==")
+			},
+			errors.New("innerOp prefix too long (52)"),
+		},
+		{
+			"failure: child size must be greater than zero",
+			func() {
+				spec.InnerSpec.ChildSize = 0
+			},
+			errors.New("spec.InnerSpec.ChildSize must be >= 1"),
+		},
+		{
+			"failure: inner op suffix malformed",
+			func() {
+				// change spec to be non-iavl spec to skip strict iavl validation
+				spec.LeafSpec.PrehashKey = HashOp_BITCOIN
```

Comment:
Don't think we need to change the hashOp in this testcase here, right?

Is this just for consistency with other testcases where prefix is modified?

Reply:
You are correct! Nice catch! I have removed it now. Ideally that hashOp change wouldn't be in the test cases, but to save some time I decided to use that trick for a few of the more difficult test cases